runtime/ts/storage/pouchdb/pouch-db-collection.ts

```diff
@@ -0,0 +1,364 @@
+        this.version++;  // yuck.
+        // TODO(lindner): fire change events here?
+      }
+
+    } catch (err) {
+      this._model = new CrdtCollectionModel();
+      this._rev = undefined;
+      // TODO(lindner): throw
+    }
+    return this._model;
+  }
+
+
+  /**
+   * Provides a way to apply changes to the model in a way that will result in the
+   * crdt being written to the underlying PouchDB.
+   * 
+   * - A new entry is stored if it doesn't exists.
+   * - If the existing entry is available it is fetched and the
+   *   internal state is updated.
+   * - A copy of the CRDT model is passed to the modelMutator, which may change it.
+   * - If the model is mutated by `modelMutator`, write a new revision and update the local
+   *   cached copy.
+   *
+   * @param modelMutator allows for modifying a copy of the underlying crdt model.
+   */
+  private async getModelAndUpdate(modelMutator: CrdtCollectionModelMutator): Promise<CrdtCollectionModel> {
+    // Keep retrying the operation until it succeeds.
+    while (1) {
+      let doc;
+      //: PouchDB.Core.IdMeta & PouchDB.Core.GetMeta & Model & {referenceMode: boolean, type: {}};
+
+      let notFound = false;
+      try {
+        doc = await this.db.get(this.pouchDbKey.location);
+        // as PouchDB.Core.IdMeta & PouchDB.Core.GetMeta & Model & {referenceMode: boolean, type: };
+
+        // Check remote doc.
+        // TODO(lindner): refactor with getModel above.
+        if (this._rev !== doc._rev) {
+          // remote revision is different, update local copy.
+          this._model = new CrdtCollectionModel(doc['model']);
+          this._rev = doc._rev;
+          this.version++;  // yuck.
+          // TODO(lindner): fire change events here?
+        }
+
+      } catch (err) {
+        if (err.name !== 'not_found') {
+          throw err;
+        }
+        notFound = true;
+        // setup basic doc, model/version updated below.
+        doc = {
+          _id: this.pouchDbKey.location,
+          referenceMode: this.referenceMode,
+          type: this.type.toLiteral()
+        };
+      }
+
+      // Run the mutator on a copy of the existing model
+      const newModel = modelMutator(new CrdtCollectionModel(this._model.toLiteral()));
+      
+      // Check if the mutator made any changes..
+      if (!notFound && JSON.stringify(this._model.toLiteral()) === JSON.stringify(newModel.toLiteral())) {
+        // mutator didn't make any changes.
+        return this._model;
+      }
+      
+      // Apply changes made by the mutator
+      doc['model'] = newModel.toLiteral();
+      doc['version'] = this.version;
+
+      // Update on pouchdb
+      try {
+        const putResult = await this.db.put(doc);
+
+        // success! update local with new model
+        this._rev = putResult.rev;
+        this._model = newModel;
+        
+        return this._model;
+      } catch (err) {
+        if (err.name === 'conflict') {
+          // keep trying;
```

Comment:
Wondered whether we could get stuck in infinite or way-too-long loop and should have some bail-out counter or at least log something periodically.

Reply:
added a TODO at the top of the while (1) to investigate this.